### PR TITLE
Uhf gnuradio interface

### PR DESCRIPTION
**Pull request by jlazzz**

### CODE REVIEW

**Comment by joshdellaz** [comment]
@Taylorr82 could you check this PR?

**Comment by joshdellaz** [comment]
@Taylorr82 could you check this PR?

File: output.bin [hunk 1]
```diff
@@ -0,0 +1 @@
+����������������~ES+W22003523 5C167331���������
```
Comment (Taylorr82) [inline output.bin:1]: Binary files should not be included

File: package.json [hunk 2]
```diff
@@ -8,11 +8,12 @@
   "private": false,
   "scripts": {
     "build:env": "sudo apt-get install libzmq3-dev -y && sudo apt install libsocketcan-dev -y && sudo apt install pkg-config -y && sudo apt install python3-dev -y && sudo apt install python3-pip -y && sudo pip3 install numpy",
-    "csp:clone": "sudo rm -rf libcsp && git clone --depth 1 https://github.com/AlbertaSat/libcsp",
-    "csp:configure": "cd libcsp && python3 waf configure --with-os=posix --enable-can-socketcan --enable-rdp --enable-crc32 --enable-hmac --enable-xtea --with-loglevel=debug --enable-debug-timestamp --enable-python3-bindings --with-driver-usart=linux --enable-if-zmqhub --enable-examples",
+    "csp:clone": " rm -rf libcsp && git clone --branch master https://github.com/Albertasat/libcsp.git",
```
Comment (Taylorr82) [inline package.json:11]: this is already covered by https://github.com/AlbertaSat/ex2_ground_station_software/pull/79

This branch should be rebased on master once that is merged

File: src/cli.py [hunk 3]
```diff
@@ -20,6 +20,7 @@
 '''  to run > sudo LD_LIBRARY_PATH=../libcsp/build PYTHONPATH=../libcsp/build python3 src/cli.py -I uart -d /dev/ttyUSB1  '''
 import time
 from groundStation import groundStation
+import traceback
```
Comment (Taylorr82) [inline src/cli.py:23]: unused include

File: src/cli.py [hunk 4]
```diff
@@ -50,6 +51,7 @@ def cli():
                     [print(key,':',value) for key, value in resp.items()]
             
         except Exception as e:
+            #traceback.print_tb()
```
Comment (Taylorr82) [inline src/cli.py:54]: Is this code needed?

File: src/groundStation/groundStation.py [hunk 5]
```diff
@@ -119,7 +131,7 @@ def __connectionManager__(self, server, port):
                 if server == 4:
                     conn = libcsp.connect(libcsp.CSP_PRIO_NORM, server, port, 1000, libcsp.CSP_O_CRC32)
                 else:
-                    conn = libcsp.connect(libcsp.CSP_PRIO_NORM, server, port, 1000, libcsp.CSP_O_RDP)
+                    conn = libcsp.connect(libcsp.CSP_PRIO_NORM, server, port, 1000000000, libcsp.CSP_O_NONE)
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:134]: already covered by https://github.com/AlbertaSat/ex2_ground_station_software/pull/79

File: src/groundStation/groundStation.py [hunk 6]
```diff
@@ -94,6 +101,11 @@ def __uart__(self, device):
         libcsp.rtable_load('1 uart, 4 uart 1')
         return ser
 
+    def __sdr__(self, device, uhf_baudrate):
+        """ Initialize SDR interface """
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:105]: covered by https://github.com/AlbertaSat/ex2_ground_station_software/pull/79

File: src/groundStation/groundStation.py [hunk 7]
```diff
@@ -161,6 +178,7 @@ def getInput(self, prompt=None, inVal=None):
         return command['dst'], command['dport'], toSend
 
     def transaction(self, server, port, buf):
+
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:181]: unnecessary whitespace

File: src/groundStation/groundStation.py [hunk 8]
```diff
@@ -316,16 +346,20 @@ def getOptions(self):
             type=int,
             default='15000', # 15 seconds
             help='RDP connection timeout')
+
+        self.parser.add_argument('-u', action='store_true')#UHF connection (not uart) enabled
         return self.parser.parse_args(sys.argv[1:])
 
 
+
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:354]: unnecesary whitespace

File: src/groundStation/uTransceiver.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                #self.listen()
+            else:
+                print('Error: Command not sent. Wait for pipe mode to expire')
+                print('Pipe mode timer set to ' + str(self.pipetimeout_s) + 's')
+        else:
+            print('UHF functionality not enabled. Please run CLI with -u flag to enable.')
+
+
+#typedef enum{
+#	U_GOOD_CONFIG =  0,
```
Comment (Taylorr82) [inline src/groundStation/uTransceiver.py:119]: Commented code should be removed. These values should be added to the python bindings though so you can access them in python

File: src/groundStation/uTransceiver.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        socket = context.socket(zmq.SUB)
+        socket.connect("tcp://localhost:%s" % port)
+        socket.setsockopt(zmq.SUBSCRIBE, b"")
+        #socket.RCVTIMEO = 100
+
+        # Initialize poll set
+        poller = zmq.Poller()
+        poller.register(socket, zmq.POLLIN)
+        
+        # s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (Taylorr82) [inline src/groundStation/uTransceiver.py:43]: Commented code

File: src/groundStation/uTransceiver.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+import sys
+import zmq
+import threading
+import os.path
+import socket
+import time
+
+
+
+# command format follows https://docs.google.com/spreadsheets/d/1zNhxhs0KJCp1187Vm3-zAzQHCY31f77l-0nlQmfXu1w/edit#gid=565953736
```
Comment (Taylorr82) [inline src/groundStation/uTransceiver.py:12]: Access denied to this. Relevant information should be placed either in a docstring here or a documentation file

**Review by Taylorr82 — CHANGES_REQUESTED** [review]

File: src/groundStation/uTransceiver.py (hunk 10 above)
Comment (joshdellaz) [reply src/groundStation/uTransceiver.py:43]: For testing purposes. Will add comment saying so.

File: src/cli.py (hunk 4 above)
Comment (joshdellaz) [reply src/cli.py:54]: For testing purposes. Added comment

File: src/cli.py (hunk 3 above)
Comment (joshdellaz) [reply src/cli.py:23]: Used for testing purposes. Added comment

File: output.bin (hunk 1 above)
Comment (joshdellaz) [reply output.bin:1]: Will remove

**Review by Taylorr82 — DISMISSED** [review]

File: package.json [hunk 12]
```diff
@@ -12,7 +12,8 @@
     "csp:configure": "cd libcsp && python3 waf configure --with-os=posix --enable-SDR --enable-rdp --enable-crc32 --enable-hmac --enable-xtea --with-loglevel=debug --enable-debug-timestamp --enable-python3-bindings --with-driver-usart=linux --enable-examples",
     "csp:build": "cd libcsp && python3 waf build",
     "csp:install": "cd libcsp &&  python3 waf install",
-    "build": "yarn csp:configure && yarn csp:build",
+    "uhf:build": "sudo rm -rf ex2_uhf_software && git clone https://github.com/AlbertaSat/ex2_uhf_software && cd ex2_uhf_software && git checkout ground_station_rebased_mods && ./build.sh",
```
Comment (Taylorr82) [inline package.json:15]: ground_station_rebased_mods has been merged. Shouldn't this just use master?

**Review by Taylorr82 — CHANGES_REQUESTED** [review]

File: package.json (hunk 12 above)
Comment (joshdellaz) [reply package.json:15]: Oh yeah, you're right... my bad

Comment (joshdellaz) [reply package.json]: Committed change

**Review by Taylorr82 — DISMISSED** [review]

File: src/groundStation/groundStation.py [hunk 13]
```diff
@@ -367,6 +391,7 @@ def getOptions(self):
     while True:
         try:
             server, port, toSend = csp.getInput(prompt='to send: ')
+            csp.handlePipeMode()
```
Comment (jmmabanta) [inline src/groundStation/groundStation.py:394]: We should probably also handle `handlePipeMode()` in cli.py (on line 36: `gs.handlePipeMode()`) so that this functionality is also available to the cli.

File: src/groundStation/uTransceiver.py (hunk 11 above)
Comment (jmmabanta) [reply src/groundStation/uTransceiver.py:11]: Would be nice to be made available so I can plan on how the website should handle these commands

**Review by jmmabanta — CHANGES_REQUESTED** [review]
Are these UHF direct commands just a prefix change to the ones already documented in CommandDocs.txt or are there more commands exclusive to UHF-Direct that we should know about (I don't have access to the document referred to in line 11 of uTransceiver.py). If there are any extra exclusive UHF-Direct commands, would it be possible if you can insert them into system.py so they can be documented in CommandDocs.txt?

The website uses that txt document to populate a database of valid commands and it would be nice if any extra commands are added into system.py. If it is just a prefix change, then no worries, it would just be a change in the website's validation logic.

Lastly, are any of these UHF-direct commands allowed to be in a command schedule? I'm assuming not but it would be nice to know so we can configure the command scheduler UI to either allow or ignore these commands.

**Comment by joshdellaz** [comment]
@jmmabanta the uhf-direct commands in question are entirely new ones and they're "special" in the sense that must be private (I'm not sure if we can even share them with people who haven't signed an NDA). Hence this somewhat odd way of doing things. These commands shouldn't need to be used regularly. The only time we'd really need to use them is if our OBC isn't working in-orbit, which would be very anomalous. Because they follow a different format and don't involve CSP, I'm not sure it would be possible (or worth the effort) to add them to system.py and have them be website-accessible.

**Comment by jmmabanta** [comment]
If that is the case then great! We can just completely ignore these from the website-side since it seems to be emergency use only, but are there any security measures planned to restrict access to these commands in ex2_ground_station_software (just in case) or is it not as critical?

File: src/groundStation/groundStation.py (hunk 13 above)
Comment (joshdellaz) [reply src/groundStation/groundStation.py:394]: Great catch, thx.

File: src/groundStation/groundStation.py [hunk 14]
```diff
@@ -290,6 +299,18 @@ def receive(self):
...
+    def handlePipeMode(self):
+        if self.uTrns_enable == True:
+            if (time.time() - self.uTrns.last_tx_time) > self.uTrns.pipetimeout_s:
+                self.uTrns.enterPipeMode()
+                #may need to add delay here?
+                command = self.parser.parseInputValue('obc.general.UHF_IS_IN_PIPE_NOTIFICATION(1)')
+                toSend = libcsp.buffer_get(len(command['args']))
+                if len(command['args']) > 0:
+                    libcsp.packet_set_data(toSend, command['args'])
+                self.transaction(command['dst'], command['dport'], toSend)
```
[the remark this answers is not in the document]
Comment (joshdellaz) [reply src/groundStation/groundStation.py:312]: Agreed

**Comment by joshdellaz** [comment]
@jmmabanta When the GS is built, these commands will only be available if people have permission to the ex2_uhf_software repo, so things should be secure.

**Review by jmmabanta — DISMISSED** [review]

**Review by jmmabanta — APPROVED** [review]
